File: Torch/LeakyReLU.lua
-- Copyright 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the license found in the
-- LICENSE file in the root directory of this source tree.

return function(sparseconvnet)
  local C = sparseconvnet.C
  local math = require 'math'
  local LeakyReLU, parent = torch.class(
    'sparseconvnet.LeakyReLU', 'nn.Module', sparseconvnet)

  function LeakyReLU:__init(leakage,ip)
    parent.__init(self)
    self.inplace = type(ip)~='boolean' or ip --default to inplace
    self.leakage = leakage

    self.output = {
      features = ip and "Recycle input.features" or torch.Tensor(),
    }
    self.gradInput = {
      features = ip and "Recycle gradOutput.features" or torch.Tensor()
    }
  end

  function LeakyReLU:updateOutput(input)
    self.output.metadata=input.metadata
    self.output.spatialSize = input.spatialSize
    C.typedFn(self._type,'LeakyReLU_updateOutput')(
      input.features:cdata(),
      self.output.features:cdata(),
      self.leakage)
    return self.output
  end

  function LeakyReLU:updateGradInput(input, gradOutput)
    if self.inplace then
      self.gradInput.features = gradOutput.features
    else
      self.gradInput.features:resizeAs(gradOutput.features)
    end
    C.typedFn(self._type,'LeakyReLU_updateGradInput')(
      input.features:cdata(),
      self.gradInput.features:cdata(),
      gradOutput.features:cdata(),
      self.leakage)
    return self.gradInput
  end

  function LeakyReLU:__tostring()
    local s = 'LeakyReLU('..self.leakage..')'
    return s
  end

  function LeakyReLU:clearState()
    self.output = {
      features = self.inplace and "Recycle input.features" or self.output.features:set(),
    }
    self.gradInput = {
      features = self.inplace and "Recycle gradOutput.features" or self.gradInput.features:set()
    }
  end

  function LeakyReLU:suggestInputSize(nOut)
    return nOut
  end
end

File: PyTorch/sparseconvnet/legacy/leakyReLU.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import torch
import sparseconvnet
from . import SparseModule
from ..utils import toLongTensor, typed_fn, optionalTensor, nullptr
from .sparseConvNetTensor import SparseConvNetTensor


class LeakyReLU(SparseModule):
    def __init__(self, leakage=0.333, ip=True):
        SparseModule.__init__(self)
        self.inplace = ip
        self.leakage = leakage
        self.output = SparseConvNetTensor(torch.Tensor())
        self.gradInput = None if ip else torch.Tensor()

    def updateOutput(self, input):
        self.output.metadata = input.metadata
        self.output.spatial_size = input.spatial_size
        typed_fn(input.features, 'LeakyReLU_updateOutput')(
            input.features,
            self.output.features,
            self.leakage)
        return self.output

    def updateGradInput(self, input, gradOutput):
        if self.inplace:
            self.gradInput = gradOutput
        typed_fn(input.features, 'LeakyReLU_updateGradInput')(
            input.features,
            self.gradInput,
            gradOutput,
            self.leakage)
        return self.gradInput

    def type(self, t, tensorCache=None):
        if t:
            self.output.type(t)
            self.gradInput = self.gradInput.type(t)


class Tanh(SparseModule):
    def __init__(self):
        SparseModule.__init__(self)
        self.output = SparseConvNetTensor(torch.Tensor())
        #self.gradInput = None if ip else torch.Tensor()
        self.gradInput = torch.Tensor()

    def updateOutput(self, input):
        self.output.metadata = input.metadata
        self.output.spatial_size = input.spatial_size
        self.output.features=torch.tanh(input.features)
        return self.output

    def updateGradInput(self, input, gradOutput):
        self.gradInput.resize_as_(gradOutput).copy_(gradOutput)
        self.gradInput.mul(1+self.output.features)
        self.gradInput.mul(1-self.output.features)
        return self.gradInput

    def type(self, t, tensorCache=None):
        if t:
            self.output.type(t)
            self.gradInput = self.gradInput.type(t)
